### PR TITLE
 Order Fulfillment: Add custom providers at the beginning of the list of shipment providers 

### PR DESCRIPTION
**Pull request by ctarda**
Implements the first part of #847 : add a "Custom" section at the beginning of the list of shipment providers.

I split #847 in two because it seems like adding the store's country providers to the beginning of the list right require (as far as I can tell) touching the Storage and Yosemite layers. Learning from past mistakes, it seems like it might be better to go merging improvements incrementally than attacking everything in a huge PR.

## Changes
All changes are isolated in `ShipmentProvidersViewController` and `ShippingProvidersViewModel`. 

As mentioned [in slack](p1556518345079200-slack-woo-mobile), I didn't manage to find a way to sort the list of providers the way it is required for this ticket if I add the "Custom" group to the CoreData store manually. 

The only way I found to be able to sort the list as required is encapsulating the data model in the view model and creating a "Custom" section on the fly.

## Testing
- Checkout the branch, run the tests
- Navigate to an order > Fulfill order > Add Tracking
- Tap the first row (Shipping Provider)

It should look like this:
Simulator Screen Shot - iPhone 8 - 2019-04-29 at 16 25 17

Notice the first section (Custom). Tapping on it should navigate to the screen to introduce a custom shipment provider. Tapping any other provider should mark it as selected.

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by mindgraffiti** [comment]
Hey @ctarda 👋 I read your :rubber_duck: thread p1556518345079200-slack-C6H8C3G23. 

I'm not used to complex / heavy data manipulation on a resultsController. This is probably going to sound naive, so thanks in advance for being patient with me and let me know if this is a bad idea.

If you have the resultsController abstracted away in the ViewModel, where the UI isn't going to grab data from it directly, why not go ahead and pull the data into an array (make it a private variable) and manipulate it however you need it? Then create another ViewModel function that returns the data to the table. I'm not sure if the resultsController can still be a lazy var if we depend on the array for manipulating the data. 

Pseudo code incoming...

```
/// At the top, where variables are declared in the ViewModel
///
private var shipmentProvidersData = []()

/// Figure out where to fire this function in the viewcontroller.
/// Needs to happen after the resultsController is ready
///
func prepareTableData() {
    tempArray = resultsController.allObjects
    var customSection 
    var countrySection
    tempArray.map {
         if $0.group.name == "Custom" ? customSection = $0 : // no-op
         if $0.group.name == "MyStoreCountry" ? $0 : // no-op
    }

    // delete the customSection and countrySection from the tempArray

    // arrange the remaining group names in alphabetical order

    // insert the customSection at index 0

    // insert the countrySection at index 0
    
} 
```

File: WooCommerce/Classes/ViewModels/ShippingProvidersViewModel.swift [hunk 1]
```diff
@@ -7,12 +7,13 @@ import Yosemite
 final class ShippingProvidersViewModel {
     let order: Order
 
+    /// Title of view displaying all available Shipment Tracking Providers
     let title = NSLocalizedString("Shipping Providers",
                                   comment: "Title of view displaying all available Shipment Tracking Providers")
 
     /// ResultsController: Surrounds us. Binds the galaxy together. And also, keeps the UITableView <> (Stored) StorageShipmentTrackingProviderGroup in sync.
     ///
-    private(set) lazy var resultsController: ResultsController<StorageShipmentTrackingProvider> = {
```
Comment (astralbodies) [inline WooCommerce/Classes/ViewModels/ShippingProvidersViewModel.swift]: Was the private setter removed intentionally?

Comment (ctarda) [reply WooCommerce/Classes/ViewModels/ShippingProvidersViewModel.swift]: Yes, to make the property fully private.

**Comment by ctarda** [comment]
👋🏼 @mindgraffiti @astralbodies 

I have pushed a couple of commits that I don't think make too much of a difference. To be honest, at this moment, I don't even understand how this screen is actually rendering any information.

The way I understood `ResultsController`, it is just a wrapper around `NSFetchedResultsController`, so, if my understanding is correct, every time we ask it to `performFetch` it should go to the coredata store and fetch whatever we ask it to do. Is that correct?

Anyway, this is what I have tried:

### Approach Number 1: Dispatching the action to load tracking providers to the previous screen
- Moved the content of the current `ShippingProvidersViewModel.fetchGroups` to `ManualTrackingViewController`. That way, the shipment tracking providers would be fetched (from the API) when we present the UI to add a shipment tracking. 
- On `viewDidLoad` of `ShipmentProvidersViewController`, ask `ShippingProvidersViewModel` to `configureResultsController` which would set the `onDidChangeContent` and `onDidResetContent` closures, and then `performFetch()` on `resultsController`
*The two closures never get called*

### Approach Number 2: Dispatching the action to load tracking providers when `ShippingProvidersViewModel` is instantiated
- The view model calls `self.fetchGroups` on its initialiser. 
- On `viewDidLoad` of `ShipmentProvidersViewController`, ask `ShippingProvidersViewModel` to `configureResultsController` which would set the `onDidChangeContent` and `onDidResetContent` closures, and then `performFetch()` on `resultsController`
*The two closures never get called*

### Approach Number 2: Dispatching the action to load tracking providers when  `ShippingProvidersViewModel` is instantiated and call `configureResultsController` after the action is completed successfully
- The view model calls `self.fetchGroups` on its initialiser. 
- The view model calls `self.configureResultsController` after the action `ShipmentAction.synchronizeShipmentTrackingProviders` is completed without an error.
*The two closures never get called*

I have seen approach 2 working only one time after cleaning the simulator. I have never been able to see it work again.

I am leaning, in general, towards approach #1 (requesting the business layer to load the tracking providers when the UI flow to add a tracking is presented), and then making the view model backing the list of tracking providers (`ShippingProvidersViewModel`) fetch the loaded providers from coredata. But, as mentioned, I can not make it work.

Any hints, suggestions, nudges in the right direction, even suggestions to change careers would be welcome at this moment

**Comment by bummytime** [comment]
@ctarda I think you are pretty close here actually. A couple of points:

1. Move `ShipmentProvidersViewController.configureViewModel()` to init() → we had a similar issue in `PeriodDataViewController` where the key was to setup the FRC immediately before anything else (https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift#L107-L108)

2. When `resultsController.performFetch()` is called, the appropriate entities will be fetched immediately — `resultsController.onDidChangeContent()` won't be called unless there is a subsequent change to the CD entities after that initial call

3. The UITableView is being populated because the initial call to `resultsController.performFetch()` already retrieved the entities. So when `func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell` is called in `ShipmentProvidersViewController` , the `viewModel.titleForCellAt(indexPath)` call is fetching this initial data from the ResultsController.

4. You may want to consider fetching the providers on `ShipmentProvidersViewController` (and just handle the situation where no data is initially returned when the DB is empty.

Check out how `siteStatsResultsController`  is used on  `PeriodDataViewController`  and maybe use that as your guide → note how `siteStatsResultsController`  is setup and that calling `performFetch()` will return data via `siteStats` even before `siteStatsResultsController.onDidChangeContent()` is fired (because of a Networking Call). It may help to put a break here (https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Dashboard/MyStore/PeriodDataViewController.swift#L16) and watch when it gets called and by what when viewing the Dash.

**Comment by ctarda** [comment]
Thanks @bummytime for making the time to take a look at this, and thanks for your thorough explanation.

So, if I understand correctly, `onDidChangeContent` will only get called when there is a change to the underlying data, not when the data is fetched for the first time. And if I also understand correctly, there is no way to actually know when the data has been fetched and it's fully available, is that correct?

**Comment by ctarda** [comment]
Hiya @astralbodies @mindgraffiti @bummytime 

I just pushed an update that:
- Moves the data fetching to `ShipmentProvidersViewController`
- Adds a spinner as a table view footer to provide some visual indication of data fetching when there is no initial data.
- Cleaned up some of the most recent changes I made.

I tried to follow @mindgraffiti 's suggestion to build a single data collection in the view model, but since it seems to me that there is no easy way to detect when the list of shipment tracking providers is fully available (specially in the case when there was some previous data in the coredata store), I kept the code that handles the "Add custom" section manually. 

This is how it looks like now. Notice the spinner on the first load (which happened after resetting the content of the simulator, so the data store is empty) and how subsequent presentations of the list are faster and do not render the spinner.

Please let me know what you think!

**Comment by bummytime** [comment]
Hiya @ctarda some answers for you: 

> So, if I understand correctly, onDidChangeContent will only get called when there is a change to the underlying data, not when the data is fetched for the first time. 

That is correct.

And if I also understand correctly, there is no way to actually know when the data has been fetched and it's fully available, is that correct?

That is technically correct, but there are ways to manage this. An example would be the state machines we use in various spots throughout the app such as [OrdersViewController](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift#L731-L786) and [NotificationsViewController](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Notifications/NotificationsViewController.swift#L749-L820).

> Also, I guess if there is no new data inserted into the coredata store, onDidChangeContent would never get triggered? 

Yep!

Also, the opposite case would be true: if the remote service updates the data in the coredata store, onDidChangeContent would get triggered, is that also correct?

That is correct! 😄

**Comment by mindgraffiti** [comment]
>Adds a spinner as a table view footer to provide some visual indication of data fetching when there is no initial data.

Nice touch! :)

**Pull request merged**